amethyst_ui/src/selection.rs

```diff
@@ -0,0 +1,268 @@
+/// Generic Type:
+/// - G: Selection Group. Used to determine which entities can be selected together at the same time.
+#[derive(Debug, Serialize, Deserialize, new)]
+pub struct Selectable<G> {
+    /// The order in which entities are selected when pressing the `Tab` key or the "go to next" input action.
+    pub order: u32,
+    #[new(default)]
+    /// A multi selection group. When multiple entities are in the same selection group, they can be selected at
+    /// the same time by holding shift or control and clicking them.
+    /// You can also select the first element, then hold shift and press the keyboard arrow keys.
+    // TODO: Holding shift + arrow keys to select more.
+    // TODO: Pressing the arrow keys could optionally be binded to change the selected ui element.
+    pub multi_select_group: Option<G>,
+    #[new(default)]
+    /// Indicates if you can select multiple entities at once without having to press the shift or control key.
+    pub auto_multi_select: bool,
+    /// Indicates if this consumes the inputs. If enabled, all inputs (except Tab) will be ignored when the component is focused.
+    /// For example, the arrow keys will not change the selected ui element.
+    /// Example usage: Ui Editable Text.
+    #[new(default)]
+    pub consumes_inputs: bool,
+}
+
+impl<G: Send + Sync + 'static> Component for Selectable<G> {
+    type Storage = FlaggedStorage<Self, DenseVecStorage<Self>>;
+}
+
+/// Component indicating that a Ui entity is currently selected.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Selected;
+
+impl Component for Selected {
+    type Storage = DenseVecStorage<Self>;
+}
+
+/// System managing the selection of entities.
+/// Reacts to `UiEvent`.
+/// Reacts to Tab and Shift+Tab.
+#[derive(Debug, Default, new)]
+pub struct SelectionKeyboardSystem<G> {
+    #[new(default)]
+    window_reader_id: Option<ReaderId<Event>>,
+    #[new(default)]
+    phantom: PhantomData<G>,
+}
+
+impl<'a, G> System<'a> for SelectionKeyboardSystem<G>
+where
+    G: Send + Sync + 'static + PartialEq,
+{
+    type SystemData = (
+        Read<'a, EventChannel<Event>>,
+        Read<'a, CachedSelectionOrder>,
+        WriteStorage<'a, Selected>,
+    );
+    fn run(&mut self, (window_events, cached, mut selecteds): Self::SystemData) {
+        /*
+		Add clicked elements + shift + ctrl status.
+		If tab or shift-tab
+			remove clicked buf
+			add replace: select higher or lower id closes to previous highest old id
+		if clicked buf isn't empty
+			if check currently highest selected multiselect group
+				// if shift && ctrl -> shift only
+				if shift
+					add multiple
+				else if ctrl || auto_multi_select
+					add single
+				else
+				    add replace
+			else
+				add replace
+		*/
+
+        // Checks if tab was pressed.
+        // TODO: Controller support/Use InputEvent in addition to keys.
+        for event in window_events.read(self.window_reader_id.as_mut().unwrap()) {
```

Comment:
I think there would be a very nice usecase for the event retrigger mechanism here, listening to and dispatching custom select/focus events. Can be a refactor later, but maybe it would be nice to add a todo so we don't forget about it.

Reply:
I don't see how the retrigger fits here. There's code logic changing components on various entities. Event retrigger is mostly to re-emit events.

Reply:
I thought it would make sense to separate the system that handles the focusing (highlighting, checking focus order, and so on) from the system that dispatches the actual event. So in the end, an `Interactable` would be a retrigger that dispatches focus events, which can be handled in isolation.

Reply:
Do you mind adding more details? I'm not sure I see how that would work.
This system checks for raw input events and emits the initial "selected" events(not there yet) and manages the `Selected` components. You can't have a retrigger manage the components like that I think. Also you have to keep in mind that there can be concurrent events, like a mouse click and the `tab` key in the same frame. I want to keep the ordering deterministic for those.

Reply:
Hm yeah you're right, it's probably better to keep this specialized. My idea would've just been to have 1 more level of indirection. Instead of this system directly handling the `selected` events, there would be a retrigger that dispatches them. But it's fine like this, was just a thought